File: jointmatrixfactorization/src/main/java/de/tu_berlin/dima/bigdata/jointmatrixfactorization/solve/Joint.java
package de.tu_berlin.dima.bigdata.jointmatrixfactorization.solve;

import org.apache.mahout.math.Vector;

import de.tu_berlin.dima.bigdata.jointmatrixfactorization.type.PactVector;
import eu.stratosphere.pact.common.stubs.Collector;
import eu.stratosphere.pact.common.stubs.MatchStub;
import eu.stratosphere.pact.common.type.PactRecord;
import eu.stratosphere.pact.common.type.base.PactFloat;
import eu.stratosphere.pact.common.type.base.PactInteger;

/**
 * input1: <userID, itemID, Rating>
 * input2: <itemID, featureVector> or <userID, featureVector>
 * 
 * output: <userID, itemID, Rating, featureVector>
 * @author titicaca
 *
 */
public class Joint extends MatchStub{

	private final PactRecord outputRecord = new PactRecord();
	private final PactVector vectorWritable = new PactVector();
	
	@Override
	public void match(PactRecord ratingRecord, PactRecord featureRecord,
			Collector<PactRecord> collector) throws Exception {

		int userID = ratingRecord.getField(0, PactInteger.class).getValue();
		int itemID = ratingRecord.getField(1, PactInteger.class).getValue();
		float rating = ratingRecord.getField(2, PactFloat.class).getValue();
		Vector featureVector = featureRecord.getField(1, PactVector.class).get();
		vectorWritable.set(featureVector);
		
		outputRecord.setField(0, new PactInteger(userID));
		outputRecord.setField(1, new PactInteger(itemID));
		outputRecord.setField(2, new PactFloat(rating));
		outputRecord.setField(3, vectorWritable);
		
		collector.collect(outputRecord);
		
	}
	
}

File: JointMatrixFactorization-Scala/src/main/scala/de/tu_berlin/dima/recommendationsystem/Joint.scala
package de.tu_berlin.dima.recommendationsystem

import eu.stratosphere.api.scala.functions._
import org.apache.mahout.math.Vector;

class Joint extends JoinFunction[(Int, Int, Float), (Int, PactVector), (Int, Int, Float, PactVector)] {
  override def apply(l: (Int, Int, Float), r: (Int, PactVector)) : (Int, Int, Float, PactVector) = {
    val userID = l._1
    val itemID = l._2
    val rating = l._3
    val featureVector : Vector = r._2.get
    val vectorWritable : PactVector = new PactVector();
    vectorWritable.set(featureVector)
    val result = (userID, itemID, rating, vectorWritable)
    result
  }
}
